pilot/pkg/xds/eds.go

```diff
@@ -159,62 +159,77 @@ func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string,
-	saUpdated := s.UpdateServiceAccount(ep, hostname)
+	newIstioEndpoints := []*model.IstioEndpoint{}
+	if features.SendUnhealthyEndpoints {
+		ep.mutex.RLock()
+		oldIstioEndpoints := ep.Shards[shard]
+		ep.mutex.RUnlock()
 
-	// For existing endpoints, we need to do full push if service accounts change.
-	if saUpdated {
-		log.Infof("Full push, service accounts changed, %v", hostname)
-		pushType = FullPush
-	}
-	// Clear the cache here. While it would likely be cleared later when we trigger a push, a race
-	// condition is introduced where an XDS response may be generated before the update, but not
-	// completed until after a response after the update. Essentially, we transition from v0 -> v1 ->
-	// v0 -> invalidate -> v1. Reverting a change we pushed violates our contract of monotonically
-	// moving forward in version. In practice, this is pretty rare and self corrects nearly
-	// immediately. However, clearing the cache here has almost no impact on cache performance as we
-	// would clear it shortly after anyways.
-	s.Cache.Clear(map[model.ConfigKey]struct{}{{
-		Kind:      gvk.ServiceEntry,
-		Name:      hostname,
-		Namespace: namespace,
-	}: {}})
-	ep.mutex.Unlock()
-	if features.SendUnhealthyEndpoints && pushType == IncrementalPush {
 		// Check if new Endpoints are ready to be pushed. This check
 		// will ensure that if a new pod comes with a non ready endpoint,
 		// we do not unnecessarily push that config to Envoy.
 		// Please note that address is not a unique key. So this may not accurately
 		// identify based on health status and push too many times - which is ok since its an optimization.
 		emap := make(map[string]*model.IstioEndpoint, len(oldIstioEndpoints))
+		// Add new endpoints only if they are ever ready once to shards
+		// so that full push does not send them from shards.
 		for _, oie := range oldIstioEndpoints {
 			emap[oie.Address] = oie
 		}
 		needPush := false
 		for _, nie := range istioEndpoints {
 			if oie, exists := emap[nie.Address]; exists {
 				// If endpoint exists already, we should push if it's health status changes.
-				needPush = oie.HealthStatus != nie.HealthStatus
-			} else {
+				if oie.HealthStatus != nie.HealthStatus {
+					needPush = true
+				}
+				newIstioEndpoints = append(newIstioEndpoints, nie)
+			} else if nie.HealthStatus == model.Healthy {
 				// If the endpoint does not exist in shards that means it is a
 				// new endpoint. Only send if it is healthy to avoid pushing endpoints
 				// that are not ready to start with.
-				needPush = nie.HealthStatus == model.Healthy
-			}
-			if needPush {
-				break
+				needPush = true
+				newIstioEndpoints = append(newIstioEndpoints, nie)
 			}
 		}
 
-		if !needPush {
+		if pushType != FullPush && !needPush {
 			log.Infof("No push, either old endpoint health status did not change or new endpoint came with unhealthy status, %v", hostname)
 			pushType = NoPush
 		}
+
 	}
 
+	ep.mutex.Lock()
```

Comment:
This is not safe AFAIK. Looks like a pretty classic race.

On L165, we read the data and drop the lock

On L203, we grab the lock again and set the data

Between that, another thread could wipe out the data, add endpoints, etc. Those would be reverted on L203